spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/GlobalKTableBinderConfiguration.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.kafka.streams;
+
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.kafka.KafkaProperties;
+import org.springframework.cloud.stream.binder.kafka.properties.KafkaBinderConfigurationProperties;
+import org.springframework.cloud.stream.binder.kafka.provisioning.KafkaTopicProvisioner;
+import org.springframework.cloud.stream.binder.kafka.streams.properties.KafkaStreamsBinderConfigurationProperties;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author Soby Chacko
+ */
+@Configuration
+public class GlobalKTableBinderConfiguration {
+
+	@Bean
+	@ConditionalOnBean(name = "outerContext")
+	public BeanFactoryPostProcessor outerContextBeanFactoryPostProcessor() {
```

Comment:
Good idea about using the `ImportBeanDefinitionRegistrar`. Issue created to track that (https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/445). 

This bean registration was originally added as part of [this issue](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/411) in the binder and [this one](https://github.com/spring-cloud/spring-cloud-stream/issues/1420) in the core. We could revisit this as part of the issue created ^^.

Reply:
Why revise later, but not in this PR? Why then we need review?
I agree that we can revise some other code, but this one is new. So, what is the point to still keep it wrong way, when we have a potential solution which could be then used for other wrong places if any?

Thanks for understanding.

Reply:
Which one are you suggesting to do in this PR - `ImportBeanDefinitionRegistrar` or the refactoring of the parent-child relationship? I think we can do the former, but if its the latter, it might be better to address that as part of a separate issue as that involves some semantic changes.

Reply:
Sorry, for confusion. Right, i mean that parent-child relationship is good, but that's really going to be much wider issue.
Right now I would just like to see an `ImportBeanDefinitionRegistrar` no any `beanFactory.getBean()` in the `BeanFactoryPostProcessor`.
You can simple check for the `outerContext` `BeanDefinition` before populating a `BeanDefinition` for your purpose.

The `beanFactory.getBean()` is very dangerous in this case. See "The early BeanFactory access syndrome": https://www.google.com/search?q=The+early+BeanFactory+access+syndrome&oq=The+early+BeanFactory+access+syndrome 😉